src/librustc_parse/parser/item.rs

```diff
@@ -1701,140 +1462,121 @@ impl<'a> Parser<'a> {
         Ok(true)
     }
 
-    fn mk_item(
+    fn mk_item<K>(
         &self,
-        span: Span,
+        lo: Span,
         ident: Ident,
-        kind: ItemKind,
+        kind: K,
         vis: Visibility,
         attrs: Vec<Attribute>,
-    ) -> P<Item> {
-        P(Item { ident, attrs, id: DUMMY_NODE_ID, kind, vis, span, tokens: None })
+    ) -> Item<K> {
+        let span = lo.to(self.prev_span);
+        Item { ident, attrs, id: DUMMY_NODE_ID, kind, vis, span, tokens: None }
     }
 }
 
 /// The parsing configuration used to parse a parameter list (see `parse_fn_params`).
 pub(super) struct ParamCfg {
-    /// Is `self` is allowed as the first parameter?
-    pub is_self_allowed: bool,
+    /// Is `self` is *semantically* allowed as the first parameter?
+    /// This is only used for diagnostics.
+    pub in_assoc_item: bool,
```

Comment:
This is better removed.
With `item` matchers in macros (which are supposed to be usable inside traits, impls and extern blocks in the future) we don't always know which kind of an item we are parsing.

The diagnostic improvement from knowing the kind look insignificant.
E.g. we can still recommend to move `self` to the first position even if `self` is not supported at all in this context, these cases of double mistakes should be very rare.

Reply:
Interesting, I was thinking people would say removing it would regress diagnostics, but I'm happy to remove it as it would simplify the code.

Reply:
Same probably applies to `AssocCtxt` in the visitor interface.
I suspect we may not always know the item's kind when visiting, so visitors that need to know it (like AST validation) will have to track it themselves.
I'll leave this question to the future though.